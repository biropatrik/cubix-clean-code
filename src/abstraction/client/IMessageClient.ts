export interface IMessageClient {
    sendNotification(message: string): void;
}