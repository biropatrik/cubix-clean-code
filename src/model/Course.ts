import { StudentAlreadyAddedException } from "../exception/StudentAlreadyAddedException";
import { Student } from "./Student";

export class Course {

    constructor(
        private name: string,
        private students: Student[] = []) { }

    public getName() {
        return this.name;
    }

    public addStudent(student: Student) {
        if (this.students.includes(student)) {
            throw new StudentAlreadyAddedException('Student already added to the course!');
        }
        this.students.push(student);
    }
}