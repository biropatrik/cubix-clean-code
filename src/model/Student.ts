import { Course } from "./Course";
import { Person } from "./Person";

export class Student extends Person {

    constructor(name: string, courses?: Course[]) {
        super(name, courses);
    }
}