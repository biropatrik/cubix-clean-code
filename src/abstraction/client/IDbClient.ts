import { Course } from "../../model/Course";
import { Student } from "../../model/Student";

export interface IDbClient {
    getCourseByName(name: string): Course | undefined;
    getAllCourses(): Course[];
    addCourse(course: Course): void;
    addStudentToCourse(student: Student, courseName: string): void;
}