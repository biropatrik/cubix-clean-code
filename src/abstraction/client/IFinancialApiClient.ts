import { Student } from "../../model/Student";

export interface IFinancialApiClient {
    getIsOrderPayed(student: Student): Promise<boolean>;
    payForCourse(student: Student, courseName: string): Promise<void>;
}