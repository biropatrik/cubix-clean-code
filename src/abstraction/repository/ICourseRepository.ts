import { Course } from "../../model/Course";
import { CourseStatistic } from "../../model/CourseStatistic";
import { Student } from "../../model/Student";

export interface ICourseRepository {
    getCourseByName(name: string): Promise<Course | undefined>;
    getAllCourses(): Promise<Course[]>;
    addCourse(course: Course): Promise<void>;
    addStudentToCourse(student: Student, courseName: string): Promise<void>;
    getCourseStatistics(courseName: string): Promise<CourseStatistic>;
}