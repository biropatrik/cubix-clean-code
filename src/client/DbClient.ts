import { IDbClient } from "../abstraction/client/IDbClient";
import { Course } from "../model/Course";
import { Student } from "../model/Student";

export class DbClient implements IDbClient {

    private courses: Course[] = [new Course('Data Analytics'), new Course('Project Management'), new Course('Java')]

    getCourseByName(name: string): Course | undefined {
        let result = undefined;
        for (const course of this.courses) {
            if (course.getName() === name) {
                result = course;
            }
        }
        return result;
    }

    getAllCourses(): Course[] {
        return this.courses;
    }

    addCourse(course: Course): void {
        this.courses.push(course);
    }

    addStudentToCourse(student: Student, courseName: string): void {
        for (const course of this.courses) {
            if (course.getName() === courseName) {
                course.addStudent(student);
            }
        }
    }
}