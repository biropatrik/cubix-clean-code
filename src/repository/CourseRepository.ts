import { IDbClient } from "../abstraction/client/IDbClient";
import { ICourseRepository } from "../abstraction/repository/ICourseRepository";
import { NetworkException } from "../exception/NetworkException";
import { UnknownException } from "../exception/UnknownException";
import { Course } from "../model/Course";
import { CourseStatistic } from "../model/CourseStatistic";
import { Student } from "../model/Student";

export class CourseRepository implements ICourseRepository {

    constructor(private dbClient: IDbClient) { }

    async getCourseByName(name: string): Promise<Course | undefined> {
        try {
            return await this.dbClient.getCourseByName(name);
        } catch (error) {
            if (error instanceof NetworkException) {
                throw error;
            }
            throw new UnknownException('Unknown error happened.');
        }
    }

    async getAllCourses(): Promise<Course[]> {
        try {
            return await this.dbClient.getAllCourses();
        } catch (error) {
            if (error instanceof NetworkException) {
                throw error;
            }
            throw new UnknownException('Unknown error happened.');
        }
    }

    async addCourse(course: Course): Promise<void> {
        try {
            await this.dbClient.addCourse(course);
        } catch (error) {
            if (error instanceof NetworkException) {
                throw error;
            }
            throw new UnknownException('Unknown error happened.');
        }
    }

    async addStudentToCourse(student: Student, courseName: string): Promise<void> {
        try {
            await this.dbClient.addStudentToCourse(student, courseName);
        } catch (error) {
            if (error instanceof NetworkException) {
                throw error;
            }
            throw new UnknownException('Unknown error happened.');
        }
    }

    async getCourseStatistics(courseName: string): Promise<CourseStatistic> {
        try {
            return await this.dbClient.getCourseStatistics(courseName);
        } catch (error) {
            if (error instanceof NetworkException) {
                throw error;
            }
            throw new UnknownException('Unknown error happened.');
        }
    }
}