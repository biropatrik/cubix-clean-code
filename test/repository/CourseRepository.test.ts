import { CourseRepository } from "../../src/repository/CourseRepository";
import { DbClient } from "../../src/client/DbClient";
import { mock, mockReset } from "jest-mock-extended";
import { Course } from "../../src/model/Course";
import { Student } from "../../src/model/Student";
import { NetworkException } from "../../src/exception/NetworkException";
import { UnknownException } from "../../src/exception/UnknownException";

describe('CourseRepository tests', () => {
    let courseRepository: CourseRepository;
    const mockDbClient = mock<DbClient>();

    beforeEach(() => {
        mockReset(mockDbClient);
        courseRepository = new CourseRepository(mockDbClient);
    })

    describe('Happy paths', () => {
        it('should return a course by name', async () => {
            // Arrange
            const courseName = 'Test';
            const course = new Course(courseName);
            mockDbClient.getCourseByName.mockReturnValue(Promise.resolve(course));

            // Act
            const result = await courseRepository.getCourseByName(courseName);

            // Assert
            expect(mockDbClient.getCourseByName).toHaveBeenCalledTimes(1);
            expect(mockDbClient.getCourseByName).toHaveBeenCalledWith(courseName);
            expect(result).toMatchSnapshot();
        })

        it('should return all courses', async () => {
            // Arrange
            const courses = [new Course('Java'), new Course('SQL'), new Course('Docker')];
            mockDbClient.getAllCourses.mockReturnValue(Promise.resolve(courses));

            // Act
            const result = await courseRepository.getAllCourses();

            // Assert
            expect(mockDbClient.getAllCourses).toHaveBeenCalledTimes(1);
            expect(mockDbClient.getAllCourses).toHaveBeenCalledWith();
            expect(result).toMatchSnapshot();
        })

        it('should add new course', async () => {
            // Arrange
            const course = new Course('Test');

            // Act
            await courseRepository.addCourse(course);

            // Assert
            expect(mockDbClient.addCourse).toHaveBeenCalledTimes(1);
            expect(mockDbClient.addCourse).toHaveBeenCalledWith(course);
        })

        it('should add new student to a course', async () => {
            // Arrange
            const courseName = 'Test';
            const student = new Student('John');

            // Act
            await courseRepository.addStudentToCourse(student, courseName);

            // Assert
            expect(mockDbClient.addStudentToCourse).toHaveBeenCalledTimes(1);
            expect(mockDbClient.addStudentToCourse).toHaveBeenCalledWith(student, courseName);
        })
    })

    describe('Error paths', () => {
        it('should throw an network exception when call getCourseByName', async () => {
            // Arrange
            const courseName = 'Test';
            const exception = new NetworkException('Network error');
            mockDbClient.getCourseByName.mockImplementation(() => { throw exception });

            // Act and assert
            await expect(() => courseRepository.getCourseByName(courseName)).rejects.toThrow(exception);
            expect(mockDbClient.getCourseByName).toHaveBeenCalledTimes(1);
            expect(mockDbClient.getCourseByName).toHaveBeenCalledWith(courseName);
        })

        it('should throw an unknown exception when call getCourseByName', async () => {
            // Arrange
            const courseName = 'Test';
            const exception = new Error('error');
            const expectedException = new UnknownException('Unknown error happened.');
            mockDbClient.getCourseByName.mockImplementation(() => { throw exception });

            // Act and assert
            await expect(() => courseRepository.getCourseByName(courseName)).rejects.toThrow(expectedException);
            expect(mockDbClient.getCourseByName).toHaveBeenCalledTimes(1);
            expect(mockDbClient.getCourseByName).toHaveBeenCalledWith(courseName);
        })

        it('should throw an network exception when call getAllCourses', async () => {
            // Arrange
            const exception = new NetworkException('Network error');
            mockDbClient.getAllCourses.mockImplementation(() => { throw exception });

            // Act and assert
            await expect(() => courseRepository.getAllCourses()).rejects.toThrow(exception);
            expect(mockDbClient.getAllCourses).toHaveBeenCalledTimes(1);
            expect(mockDbClient.getAllCourses).toHaveBeenCalledWith();
        })

        it('should throw an unknown exception when call getAllCourses', async () => {
            // Arrange
            const exception = new Error('error');
            const expectedException = new UnknownException('Unknown error happened.');
            mockDbClient.getAllCourses.mockImplementation(() => { throw exception });

            // Act and assert
            await expect(() => courseRepository.getAllCourses()).rejects.toThrow(expectedException);
            expect(mockDbClient.getAllCourses).toHaveBeenCalledTimes(1);
            expect(mockDbClient.getAllCourses).toHaveBeenCalledWith();
        })

        it('should throw an network exception when call addCourse', async () => {
            // Arrange
            const course = new Course('Test');
            const exception = new NetworkException('Network error');
            mockDbClient.addCourse.mockImplementation(() => { throw exception });

            // Act and assert
            await expect(() => courseRepository.addCourse(course)).rejects.toThrow(exception);
            expect(mockDbClient.addCourse).toHaveBeenCalledTimes(1);
            expect(mockDbClient.addCourse).toHaveBeenCalledWith(course);
        })

        it('should throw an unknown exception when call addCourse', async () => {
            // Arrange
            const course = new Course('Test');
            const exception = new Error('error');
            const expectedException = new UnknownException('Unknown error happened.');
            mockDbClient.addCourse.mockImplementation(() => { throw exception });

            // Act and assert
            await expect(() => courseRepository.addCourse(course)).rejects.toThrow(expectedException);
            expect(mockDbClient.addCourse).toHaveBeenCalledTimes(1);
            expect(mockDbClient.addCourse).toHaveBeenCalledWith(course);
        })

        it('should throw an network exception when call addStudentToCourse', async () => {
            // Arrange
            const courseName = 'Test';
            const student = new Student('John');
            const exception = new NetworkException('Network error');
            mockDbClient.addStudentToCourse.mockImplementation(() => { throw exception });

            // Act and assert
            await expect(() => courseRepository.addStudentToCourse(student, courseName)).rejects.toThrow(exception);
            expect(mockDbClient.addStudentToCourse).toHaveBeenCalledTimes(1);
            expect(mockDbClient.addStudentToCourse).toHaveBeenCalledWith(student, courseName);
        })

        it('should throw an unknown exception when call addStudentToCourse', async () => {
            // Arrange
            const courseName = 'Test';
            const student = new Student('John');
            const exception = new Error('error');
            const expectedException = new UnknownException('Unknown error happened.');
            mockDbClient.addStudentToCourse.mockImplementation(() => { throw exception });

            // Act and assert
            await expect(() => courseRepository.addStudentToCourse(student, courseName)).rejects.toThrow(expectedException);
            expect(mockDbClient.addStudentToCourse).toHaveBeenCalledTimes(1);
            expect(mockDbClient.addStudentToCourse).toHaveBeenCalledWith(student, courseName);
        })
    })
})