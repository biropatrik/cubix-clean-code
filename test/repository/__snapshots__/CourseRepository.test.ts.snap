// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CourseRepository tests Happy paths should return a course by name 1`] = `
Course {
  "name": "Test",
  "students": [],
}
`;

exports[`CourseRepository tests Happy paths should return all courses 1`] = `
[
  Course {
    "name": "Java",
    "students": [],
  },
  Course {
    "name": "SQL",
    "students": [],
  },
  Course {
    "name": "Docker",
    "students": [],
  },
]
`;
